File: sdks/javascript/examples/workflow-orchestrator.ts
/**
 * Distributed Workflow Orchestrator Example (TypeScript)
 *
 * Demonstrates:
 * - Custom state machine with transitions map
 * - Actor-based saga steps with compensation
 * - Human task integration (simulated review)
 * - StateHandle for persistence / audit trail
 *
 * Since the JS SDK does not have dedicated workflow modules, this example
 * builds a simplified workflow orchestrator using the core Actor, StateHandle,
 * and Messaging primitives.
 */

import { Actor } from '../src/actor';
import { Message, MessageType } from '../src/messaging';
import { StateHandle } from '../src/state';

// ========================================
// Types
// ========================================

interface WorkflowState {
  currentState: string;
  input: Record<string, any>;
  history: AuditEntry[];
  variables: Record<string, any>;
  startedAt: string;
  updatedAt: string;
}

interface AuditEntry {
  type: string;
  timestamp: string;
  details: Record<string, any>;
}

interface SagaStepDefinition {
  name: string;
  action: (ctx: SagaContext) => Promise<any>;
  compensate?: (ctx: SagaContext) => Promise<void>;
}

interface SagaContext {
  input: Record<string, any>;
  state: Record<string, any>;
  completedSteps: string[];
  failedStep?: string;
  error?: string;
  set(key: string, value: any): void;
  get(key: string): any;
  markCompleted(step: string): void;
}

interface SagaResult {
  status: 'completed' | 'compensated' | 'failed';
  completedSteps: string[];
  compensatedSteps: string[];
  error?: string;
  durationMs: number;
}

interface HumanTaskDef {
  taskId: string;
  taskType: string;
  title: string;
  assignee: string;
  status: 'pending' | 'in_progress' | 'completed' | 'rejected';
  result?: Record<string, any>;
  completedBy?: string;
}

// ========================================
// Utilities
// ========================================

function uuid(): string {
  return Math.random().toString(36).substring(2, 10);
}

const AUDIT_LOG: string[] = [];

function log(msg: string): void {
  const ts = new Date().toISOString().substring(11, 23);
  const entry = `  [${ts}] ${msg}`;
  AUDIT_LOG.push(entry);
  console.log(entry);
}

function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

// ========================================
// State Machine
// ========================================

interface TransitionDef {
  name: string;
  from: string;
  to: string;
}

class StateMachine {
  private transitions: Map<string, TransitionDef[]> = new Map();

  addTransition(transition: TransitionDef): this {
    const fromMap = this.transitions.get(transition.from) ?? [];
    fromMap.push(transition);
    this.transitions.set(transition.from, fromMap);
    return this;
  }

  getAvailableTransitions(currentState: string): string[] {
    return (this.transitions.get(currentState) ?? []).map(t => t.name);
  }

  execute(transitionName: string, currentState: string): string | null {
    const candidates = this.transitions.get(currentState) ?? [];
    const transition = candidates.find(t => t.name === transitionName);
    return transition?.to ?? null;
  }
}

// ========================================
// Saga Executor
// ========================================

class SagaExecutor {
  async execute(
    steps: SagaStepDefinition[],
    input: Record<string, any>,
  ): Promise<SagaResult> {
    const startTime = Date.now();
    const ctx: SagaContext = {
      input,
      state: {},
      completedSteps: [],
      set(key, value) { this.state[key] = value; },
      get(key) { return this.state[key]; },
      markCompleted(step) {
        if (!this.completedSteps.includes(step)) this.completedSteps.push(step);
      },
    };

    try {
      for (const step of steps) {
        log(`  [SAGA] Executing step: ${step.name}`);
        await step.action(ctx);
        ctx.markCompleted(step.name);
        log(`  [SAGA] Step completed: ${step.name}`);
      }

      return {
        status: 'completed',
        completedSteps: ctx.completedSteps,
        compensatedSteps: [],
        durationMs: Date.now() - startTime,
      };
    } catch (err: any) {
      ctx.failedStep = steps.find(s => !ctx.completedSteps.includes(s.name))?.name;
      ctx.error = err?.message ?? String(err);
      log(`  [SAGA] Step failed: ${ctx.failedStep}: ${ctx.error}`);

      const completedReversed = [...ctx.completedSteps].reverse();
      const compensatedSteps: string[] = [];

      for (const stepName of completedReversed) {
        const step = steps.find(s => s.name === stepName);
        if (step?.compensate) {
          log(`  [COMPENSATE] Compensating: ${stepName}`);
          try {
            await step.compensate(ctx);
            compensatedSteps.push(stepName);
            log(`  [COMPENSATE] Compensated: ${stepName}`);
          } catch (compErr: any) {
            log(`  [COMPENSATE] Compensation failed for ${stepName}: ${compErr.message}`);
          }
        }
      }

      return {
        status: compensatedSteps.length > 0 ? 'compensated' : 'failed',
        completedSteps: ctx.completedSteps,
        compensatedSteps,
        error: ctx.error,
        durationMs: Date.now() - startTime,
      };
    }
  }
}

// ========================================
// Workflow Executor
// ========================================

class WorkflowExecutor {
  private stateMachine = new StateMachine();
  private workflows = new Map<string, WorkflowState>();
  private humanTasks = new Map<string, HumanTaskDef>();

  addTransition(def: TransitionDef): this {
    this.stateMachine.addTransition(def);
    return this;
  }

  async start(
    input: Record<string, any>,
    initialState: string,
  ): Promise<string> {
    const wfId = uuid();
    this.workflows.set(wfId, {
      currentState: initialState,
      input,
      history: [],
      variables: {},
      startedAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    });
    this.addAudit(wfId, 'workflow_started', { state: initialState });
    log(`Workflow started: ${wfId}, state=${initialState}`);
    return wfId;
  }

  async transition(wfId: string, transitionName: string): Promise<string | null> {
    const wf = this.workflows.get(wfId);
    if (!wf) throw new Error(`Unknown workflow: ${wfId}`);

    const newState = this.stateMachine.execute(transitionName, wf.currentState);
    if (newState === null) {
      log(`Invalid transition '${transitionName}' from '${wf.currentState}'`);
      return null;
    }

    const oldState = wf.currentState;
    wf.currentState = newState;
    wf.updatedAt = new Date().toISOString();
    this.addAudit(wfId, 'transition', { transition: transitionName, from: oldState, to: newState });
    log(`Transition: ${oldState} -> ${newState} (via ${transitionName})`);
    return newState;
  }

  getStatus(wfId: string): WorkflowState | undefined {
    return this.workflows.get(wfId);
  }

  createHumanTask(
    taskType: string,
    title: string,
    assignee: string,
    workflowId: string,
  ): HumanTaskDef {
    const task: HumanTaskDef = {
      taskId: uuid(),
      taskType,
      title,
      assignee,
      status: 'pending',
    };
    this.humanTasks.set(task.taskId, task);
    log(`Human task created: ${task.taskId} (${taskType}) assigned to ${assignee}`);
    return task;
  }

  async completeHumanTask(taskId: string, result: Record<string, any>): Promise<HumanTaskDef> {
    const task = this.humanTasks.get(taskId);
    if (!task) throw new Error(`Task not found: ${taskId}`);
    task.status = 'completed';
    task.result = result;
    task.completedBy = task.assignee;
    log(`Human task completed: ${taskId} by ${task.completedBy} => ${JSON.stringify(result)}`);
    return task;
  }

  private addAudit(wfId: string, type: string, details: Record<string, any>): void {
    const wf = this.workflows.get(wfId);
    if (wf) {
      wf.history.push({
        type,
        timestamp: new Date().toISOString(),
        details,
      });
    }
  }
}

// ========================================
// Main
// ========================================

async function runWorkflowOrchestrator(): Promise<void> {
  console.log('='.repeat(70));
  console.log('  AETHER WORKFLOW ORCHESTRATOR - Document Approval Example (TS)');
  console.log('='.repeat(70));
  console.log();

  const executor = new WorkflowExecutor();
  const sagaExec = new SagaExecutor();

  console.log('--- Step 1: Define the approval state machine ---');
  executor
    .addTransition({ name: 'submit', from: 'draft', to: 'pending_review' })
    .addTransition({ name: 'approve', from: 'pending_review', to: 'approved' })
    .addTransition({ name: 'reject', from: 'pending_review', to: 'rejected' })
    .addTransition({ name: 'saga_success', from: 'approved', to: 'completed' })
    .addTransition({ name: 'saga_failed', from: 'approved', to: 'failed' });
  console.log('  States: draft, pending_review, approved, rejected, completed, failed');
  console.log('  Transitions defined successfully');
  console.log();

  console.log('--- Step 2: Define the post-approval saga ---');

  const sagaSteps: SagaStepDefinition[] = [
    {
      name: 'notify-team',
      action: async (ctx) => {
        log('  [SAGA] Notifying team about approved document...');
        await sleep(50);
        ctx.set('notification_sent', true);
        log('  [SAGA] Team notification sent');
        return 'ok';
      },
      compensate: async (ctx) => {
        log('  [COMPENSATE] Withdrawing team notification...');
        ctx.set('notification_sent', false);
        log('  [COMPENSATE] Notification withdrawn');
      },
    },
    {
      name: 'archive-document',
      action: async (ctx) => {
        log('  [SAGA] Archiving document to storage...');
        await sleep(50);
        if (ctx.input.simulateFailure) {
          throw new Error('Storage service unavailable during archival');
        }
        ctx.set('archived', true);
        log('  [SAGA] Document archived');
        return 'ok';
      },
      compensate: async (ctx) => {
        log('  [COMPENSATE] Un-archiving document...');
        ctx.set('archived', false);
        log('  [COMPENSATE] Document un-archived');
      },
    },
    {
      name: 'send-confirmation',
      action: async (ctx) => {
        log('  [SAGA] Sending confirmation email...');
        await sleep(50);
        ctx.set('confirmation_sent', true);
        log('  [SAGA] Confirmation sent');
        return 'ok';
      },
      compensate: async (ctx) => {
        log('  [COMPENSATE] Withdrawing confirmation...');
        ctx.set('confirmation_sent', false);
        log('  [COMPENSATE] Confirmation withdrawn');
      },
    },
  ];
  console.log(`  Saga: ${sagaSteps.length} steps with compensation`);
  sagaSteps.forEach(s => {
    console.log(`    - ${s.name} (compensation: ${s.compensate ? 'yes' : 'no'})`);
  });
  console.log();

  for (const [runNum, shouldFail] of [[1, false], [2, true]] as [number, boolean][]) {
    const docTitle = `Q4 Report ${runNum}${shouldFail ? ' (failure scenario)' : ''}`;
    console.log('='.repeat(70));
    console.log(`  RUN ${runNum}: Document '${docTitle}'`);
    console.log('='.repeat(70));

    console.log('\n  >> Starting workflow...');
    const wfId = await executor.start({ title: docTitle, simulateFailure: shouldFail }, 'draft');

    console.log('\n  >> Transition: draft -> pending_review');
    await executor.transition(wfId, 'submit');

    console.log('\n  >> Creating human review task...');
    const task = executor.createHumanTask(
      'document_review',
      `Review: ${docTitle}`,
      'reviewer@company.com',
      wfId,
    );

    await sleep(30);

    console.log('\n  >> Human reviewer approves the document...');
    await executor.completeHumanTask(task.taskId, {
      approved: true,
      comments: 'Looks good, approved.',
    });

    console.log('\n  >> Transition: pending_review -> approved');
    await executor.transition(wfId, 'approve');

    console.log('\n  >> Executing post-approval saga...');
    const sagaResult = await sagaExec.execute(sagaSteps, executor.getStatus(wfId)!.input);
    log(`Saga result: status=${sagaResult.status}`);
    log(`  Completed steps: [${sagaResult.completedSteps.join(', ')}]`);
    if (sagaResult.compensatedSteps.length > 0) {
      log(`  Compensated steps: [${sagaResult.compensatedSteps.join(', ')}]`);
    }
    if (sagaResult.error) {
      log(`  Error: ${sagaResult.error}`);
    }
    log(`  Duration: ${sagaResult.durationMs}ms`);

    if (sagaResult.status === 'completed') {
      console.log('\n  >> Transition: approved -> completed');
      await executor.transition(wfId, 'saga_success');
    } else {
      console.log('\n  >> Transition: approved -> failed');
      await executor.transition(wfId, 'saga_failed');
    }

    const finalState = executor.getStatus(wfId)!;
    log(`Final state: ${finalState.currentState}`);

    console.log('\n  >> Audit trail:');
    for (const entry of finalState.history) {
      const detailStr = Object.entries(entry.details)
        .map(([k, v]) => `${k}=${v}`)
        .join(', ');
      log(`  AUDIT: ${entry.type} (${detailStr})`);
    }
    console.log();
  }

  console.log('='.repeat(70));
  console.log('  FULL AUDIT LOG');
  console.log('='.repeat(70));
  for (const entry of AUDIT_LOG) {
    console.log(entry);
  }
  console.log(`\nTotal audit entries: ${AUDIT_LOG.length}`);
}

(async () => {
  await runWorkflowOrchestrator();
})();

File: sdks/python/examples/workflow_orchestrator.py
"""
Distributed Workflow Orchestrator Example

Demonstrates:
- State machine for document approval workflow
- Saga pattern for multi-step post-approval processing
- Human task integration (review step)
- Event sourcing for full audit trail

Workflow:
    Draft -> Pending Review -> [Human Approval] -> Approved -> [Saga] -> Completed
                              -> [Human Rejection]  -> Rejected
    Saga steps: Notify Team -> Archive Document -> Send Confirmation
    If saga fails -> Compensation: Undo each completed step in reverse
"""

import asyncio
import uuid
from datetime import datetime

from aether_sdk.workflow.human_task import HumanTask, HumanTaskManager, TaskForm
from aether_sdk.workflow.saga import Saga, SagaExecutor
from aether_sdk.workflow.state_machine import Workflow, WorkflowExecutor
from aether_sdk.workflow.types import Duration, RetryConfig, RetryPolicy, SagaStatus

AUDIT_LOG = []


def log_event(msg: str):
    ts = datetime.utcnow().strftime("%H:%M:%S.%f")[:-3]
    AUDIT_LOG.append(f"[{ts}] {msg}")
    print(f"  [{ts}] {msg}")


async def run_approval_flow(simulate_saga_failure: bool = False):
    print("=" * 70)
    print("  AETHER WORKFLOW ORCHESTRATOR - Document Approval Example")
    print("=" * 70)
    print()

    human_task_manager = HumanTaskManager()

    print("--- Step 1: Define the approval state machine ---")
    approval_workflow = (
        Workflow("document-approval")
        .state("draft", is_initial=True)
        .state("pending_review")
        .state("approved")
        .state("rejected", is_final=True)
        .state("completed", is_final=True)
        .state("failed", is_final=True)
        .transition("submit", "draft", "pending_review")
        .transition("approve", "pending_review", "approved")
        .transition("reject", "pending_review", "rejected")
        .transition("saga_success", "approved", "completed")
        .transition("saga_failed", "approved", "failed")
        .on_enter(
            "pending_review",
            lambda ctx: log_event(
                f"ENTER pending_review: Document '{ctx.input['title']}' sent for review"
            ),
        )
        .on_enter(
            "approved",
            lambda ctx: log_event(
                "ENTER approved: Document has been approved by reviewer"
            ),
        )
        .on_enter(
            "rejected",
            lambda ctx: log_event("ENTER rejected: Document has been rejected"),
        )
        .on_enter(
            "completed",
            lambda ctx: log_event(
                "ENTER completed: All post-approval steps finished successfully"
            ),
        )
        .on_enter(
            "failed",
            lambda ctx: log_event(
                "ENTER failed: Post-approval saga failed, compensation applied"
            ),
        )
        .build()
    )
    print(f"  States: {list(approval_workflow.states.keys())}")
    print(f"  Initial: {approval_workflow.initial_state}")
    print()

    print("--- Step 2: Define the post-approval saga ---")

    async def notify_team(ctx):
        log_event("  [SAGA] Notifying team about approved document...")
        await asyncio.sleep(0.1)
        ctx.set_state("notification_sent", True)
        log_event("  [SAGA] Team notification sent successfully")
        return "notification_ok"

    async def undo_notify_team(ctx):
        log_event("  [COMPENSATE] Withdrawing team notification...")
        ctx.set_state("notification_sent", False)
        log_event("  [COMPENSATE] Team notification withdrawn")

    async def archive_document(ctx):
        log_event("  [SAGA] Archiving document to storage...")
        await asyncio.sleep(0.1)
        if simulate_saga_failure and not ctx.is_step_completed("archive-document"):
            raise RuntimeError("Storage service unavailable during archival")
        ctx.set_state("archived", True)
        log_event("  [SAGA] Document archived successfully")
        return "archive_ok"

    async def undo_archive_document(ctx):
        log_event("  [COMPENSATE] Un-archiving document from storage...")
        ctx.set_state("archived", False)
        log_event("  [COMPENSATE] Document un-archived")

    async def send_confirmation(ctx):
        log_event("  [SAGA] Sending confirmation to submitter...")
        await asyncio.sleep(0.1)
        ctx.set_state("confirmation_sent", True)
        log_event("  [SAGA] Confirmation sent successfully")
        return "confirmation_ok"

    async def undo_send_confirmation(ctx):
        log_event("  [COMPENSATE] Withdrawing confirmation...")
        ctx.set_state("confirmation_sent", False)
        log_event("  [COMPENSATE] Confirmation withdrawn")

    approval_saga = (
        Saga("post-approval")
        .step("notify-team")
        .action(notify_team)
        .compensate(undo_notify_team)
        .step("archive-document")
        .action(archive_document)
        .compensate(undo_archive_document)
        .retry(
            RetryConfig(
                max_attempts=2,
                policy=RetryPolicy.FIXED,
                initial_delay=Duration.from_seconds(0.1),
            )
        )
        .step("send-confirmation")
        .action(send_confirmation)
        .compensate(undo_send_confirmation)
        .build()
    )
    print(f"  Saga: {approval_saga.name} ({len(approval_saga.steps)} steps)")
    for step in approval_saga.steps:
        has_comp = step.compensate is not None
        print(f"    - {step.name} (compensation: {'yes' if has_comp else 'no'})")
    print()

    print("--- Step 3: Create the review task form ---")
    review_form = TaskForm()
    review_form.add_field("approved", "boolean", required=True, label="Approve?")
    review_form.add_field("comments", "text", label="Review Comments")
    print(f"  Form fields: {[f.name for f in review_form.fields]}")
    print()

    workflow_executor = WorkflowExecutor()
    saga_executor = SagaExecutor()

    for run_num, should_fail in enumerate([False, simulate_saga_failure], 1):
        doc_title = f"Q4 Report {run_num} {'(failure scenario)' if should_fail else ''}"
        print(f"{'=' * 70}")
        print(f"  RUN {run_num}: Document '{doc_title}'")
        print(f"{'=' * 70}")

        doc_id = str(uuid.uuid4())[:8]
        log_event(f"Submitting document '{doc_title}' (id={doc_id})")

        print("\n  >> Transition: draft -> pending_review")
        result = await workflow_executor.start(
            approval_workflow,
            {"title": doc_title, "id": doc_id},
        )
        wf_id = result.workflow_id
        log_event(f"Workflow started: {wf_id}, state={result.current_state}")

        await workflow_executor.transition(wf_id, "submit")
        status = await workflow_executor.get_status(wf_id)
        log_event(f"Current state: {status.current_state}")

        print("\n  >> Creating human review task...")
        task = (
            HumanTask(
                task_type="document_review",
                title=f"Review: {doc_title}",
                description=f"Please review document {doc_id}",
            )
            .with_assignee("reviewer@company.com")
            .with_priority(2)
            .with_form(review_form)
        )
        created_task = await human_task_manager.create_task(task, wf_id, "review-step")
        log_event(f"Review task created: {created_task.task_id}")
        log_event(f"  Assigned to: {created_task.assignee}")

        await asyncio.sleep(0.05)

        print("\n  >> Human reviewer approves the document...")
        await human_task_manager.claim_task(
            created_task.task_id, "reviewer@company.com"
        )
        await human_task_manager.complete_task(
            created_task.task_id,
            {"approved": True, "comments": "Looks good, approved."},
            user="reviewer@company.com",
        )
        log_event(f"Task completed by {created_task.completed_by}")
        log_event(f"  Result: {created_task.result}")

        print("\n  >> Transition: pending_review -> approved")
        await workflow_executor.transition(wf_id, "approve")
        status = await workflow_executor.get_status(wf_id)
        log_event(f"Current state: {status.current_state}")

        print("\n  >> Executing post-approval saga...")
        saga_result = await saga_executor.execute(
            approval_saga,
            {"title": doc_title, "id": doc_id, "workflow_id": wf_id},
        )
        log_event(f"Saga result: status={saga_result.status.value}")
        log_event(f"  Completed steps: {saga_result.completed_steps}")
        if saga_result.compensated_steps:
            log_event(f"  Compensated steps: {saga_result.compensated_steps}")
        if saga_result.error:
            log_event(f"  Error: {saga_result.error}")
        if saga_result.duration_ms:
            log_event(f"  Duration: {saga_result.duration_ms}ms")

        if saga_result.status == SagaStatus.COMPLETED:
            print("\n  >> Transition: approved -> completed")
            await workflow_executor.transition(wf_id, "saga_success")
        else:
            print("\n  >> Transition: approved -> failed")
            await workflow_executor.transition(wf_id, "saga_failed")

        final_status = await workflow_executor.get_status(wf_id)
        log_event(
            f"Final state: {final_status.current_state} (status={final_status.status.value})"
        )

        print(f"\n  >> Audit trail for workflow {wf_id}:")
        for event in final_status.history:
            detail = event.get("details", {})
            _ = event.get("timestamp", "?")
            evt_type = event.get("type", "?")
            detail_str = ", ".join(f"{k}={v}" for k, v in detail.items())
            log_event(f"  AUDIT: {evt_type} ({detail_str})")

        print()

    print("=" * 70)
    print("  FULL AUDIT LOG")
    print("=" * 70)
    for entry in AUDIT_LOG:
        print(entry)
    print(f"\nTotal audit entries: {len(AUDIT_LOG)}")


if __name__ == "__main__":
    asyncio.run(run_approval_flow(simulate_saga_failure=True))
